### PR TITLE
Exempt borrowing implementation

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR is intended to implement a recent enhancement to https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness#dispatching , borrowing by the exempt priority level.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
@tkashem started the work on this, in #117703 .  He has gone on vacation and I am picking up the work.
This PR is not complete yet.  It still needs to (a) change the config-consuming controller to connect the new configuration possibility with the corresponding new behavior in QueueSet and (b) finish adjusting the config-producing controller.

#### Does this PR introduce a user-facing change?

```release-note
In the API Priority and Fairness feature, priority levels that are exempt from limitation can now be given a nominal and a lendable concurrency and their dispatching borrows from the concurrency limits of the other priority levels.  For details see https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness#dispatching .
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP] https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness
```

/cc @wojtek-t 
/cc @deads2k 
@tkashem 
@lavalamp

### CODE REVIEW

**Comment by jiahuif** [comment]
/assign @wojtek-t 
Could you take a look at this? Thank you.
/api-review
/triage accepted

**Comment by MikeSpreitzer** [comment]
The force-push to 45156e107fd is a rebase onto `master`.

**Comment by MikeSpreitzer** [comment]
The explanation of `AutoUpdateAnnotationKey = "apf.kubernetes.io/autoupdate-spec"` needs to be updated due to the exception carved out for the exempt PriorityLevelConfiguration.

**Comment by MikeSpreitzer** [comment]
This is also relevant: it shows that for most of the time, priority levels could get about twice what they wanted.

Screen Shot 2023-06-26 at 1 24 31 PM

**Comment by MikeSpreitzer** [comment]
However, there is this.  The one that saturated at the end is `exempt`.

Screen Shot 2023-06-26 at 1 33 36 PM

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 1]
```diff
@@ -197,16 +197,17 @@ type priorityLevelState struct {
 	pl *flowcontrol.PriorityLevelConfiguration
 
 	// qsCompleter holds the QueueSetCompleter derived from `config`
-	// and `queues` if config is not exempt, nil otherwise.
+	// and `queues`.
+	// Formerly: non-nil if config is not exempt, nil otherwise.
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:201]: I wouldn't add comments about what it formerly was - it's overcomplicating it imho.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 2]
```diff
@@ -384,7 +385,7 @@ func (cfgCtlr *configController) updateBorrowingLocked(setCompleters bool, plSta
 	items := make([]allocProblemItem, 0, len(plStates))
 	plNames := make([]string, 0, len(plStates))
 	for plName, plState := range plStates {
-		if plState.pl.Spec.Limited == nil {
+		if false && plState.pl.Spec.Limited == nil {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:388]: ??

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 3]
```diff
@@ -412,15 +413,15 @@ func (cfgCtlr *configController) updateBorrowingLocked(setCompleters bool, plSta
 		allocs = make([]float64, len(items))
 		for idx, plName := range plNames {
 			plState := plStates[plName]
-			if plState.pl.Spec.Limited == nil {
+			if false && plState.pl.Spec.Limited == nil {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:416]: ?? [same in some places below]

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 4]
```diff
@@ -441,8 +442,15 @@ func (cfgCtlr *configController) updateBorrowingLocked(setCompleters bool, plSta
 		if relChange >= 0.05 {
 			logLevel = 2
 		}
-		klog.V(logLevel).InfoS("Update CurrentCL", "plName", plName, "seatDemandHighWatermark", plState.seatDemandStats.highWatermark, "seatDemandAvg", plState.seatDemandStats.avg, "seatDemandStdev", plState.seatDemandStats.stdDev, "seatDemandSmoothed", plState.seatDemandStats.smoothed, "fairFrac", fairFrac, "currentCL", currentCL, "backstop", err != nil)
-		plState.queues = plState.qsCompleter.Complete(fq.DispatchingConfig{ConcurrencyLimit: currentCL})
+		var concurrencyDenominator int
+		if currentCL > 0 {
+			concurrencyDenominator = currentCL
+		} else {
+			concurrencyDenominator = int(math.Max(1, math.Round(float64(cfgCtlr.serverConcurrencyLimit)/10)))
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:449]: Can you please remind me, why we're using "serverConcurrencyLimit/10"? [where is "/10" coming from]?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go [hunk 5]
```diff
@@ -809,82 +815,87 @@ func (meal *cfgMeal) processOldPLsLocked() {
...
 func (meal *cfgMeal) finishQueueSetReconfigsLocked() {
 	for plName, plState := range meal.newPLStates {
-		if plState.pl.Spec.Limited == nil {
+		if false && plState.pl.Spec.Limited == nil {
 			klog.V(5).Infof("Using exempt priority level %q: quiescing=%v", plName, plState.quiescing)
 			continue
 		}
-
-		limited := plState.pl.Spec.Limited
+		nominalConcurrencyShares, lendablePercent, borrowingLimitPercent := plSpecCommons(plState.pl)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:822]: I'm wondering if we can move these refactorings to a separate PR to submit them faster and make the actual implementation PR simpler thanks to it.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go (hunk 4 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:449]: It's just a number I made up.  I am open to suggestions for what would be better.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:201]: I left some comments about "formerly" and `false &&` as breadcrumbs while developing/debugging this PR.  They will be removed.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go (hunk 5 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go:822]: So the idea is to first do a PR that makes QueueSet support exempt behavior and makes the config-consuming controller use that, then follow with this PR that adds borrowing by exempt?

**Comment by wojtek-t** [comment]
@MikeSpreitzer - please rebase

**Comment by MikeSpreitzer** [comment]
The force-push to 7ef3d6e2ae3 is a rebase onto `master`.

**Comment by MikeSpreitzer** [comment]
I looked at the Prometheus metrics from the latest run of pull-kubernetes-e2e-gce-100-performance.

These two graphs show the result of and an intermediate variable in the borrowing logic.

Screen Shot 2023-07-07 at 12 43 52 AM

Screen Shot 2023-07-07 at 12 45 00 AM

**Comment by MikeSpreitzer** [comment]
This one shows the actual seat occupancy, grossly undersampled (as usual).

Screen Shot 2023-07-07 at 12 46 51 AM

**Comment by MikeSpreitzer** [comment]
This one shows the 90th percentile of the relative seat utilization observed every nanosecond.  For the `exempt` priority level the denominator hits the special case of configured zero and so 60 (1/10 of the server's total capacity) was used instead.  The `exempt` priority level hugs the horizontal axis.

Screen Shot 2023-07-07 at 12 52 58 AM

**Comment by MikeSpreitzer** [comment]
This one is 99th percentile of relative seat occupancy observed every nanosecond.  The `exempt` priority level is highlighted.  It gets up to 1.0 near the end of the run.

Screen Shot 2023-07-07 at 12 54 50 AM

**Comment by MikeSpreitzer** [comment]
The force-push to 29dd83a13ed updates the comment on `AutoUpdateAnnotationKey = "apf.kubernetes.io/autoupdate-spec"`.

**Comment by MikeSpreitzer** [comment]
The pull-kubernetes-integration CI job seems to be consistently failing.  #119216

File: pkg/apis/flowcontrol/types.go [hunk 6]
```diff
@@ -447,6 +455,43 @@ type LimitedPriorityLevelConfiguration struct {
...
+	// priority levels in proportion to their NCS values:
+	//
+	// NominalCL(i)  = ceil( ServerCL * NCS(i) / sum_ncs )
+	// sum_ncs = sum[priority level k] NCS(k)
+	//
+	// Bigger numbers mean a larger nominal concurrency limit,
+	// at the expense of every other priority level.
+	// This field has a default value of zero.
+	// +optional
+	NominalConcurrencyShares int32
```
Comment (deads2k) [inline pkg/apis/flowcontrol/types.go:478]: So if I'm understanding correctly, this field will be used so that if an exempt priority level uses more than this number of concurrency shares, it will start borrowing from other priority levels.  This is done so that an exempt priority level using greater and greater requests will consume the available concurrency from limited priority levels to keep the overall concurrency at a predictable level.

Even if this priority level runs out of others to borrow from, requests will still be dispatched.

If I've understood correctly, then I need to check the validation and defaulting rules and then I think the API aspect is ok.

Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/types.go:478]: Primarily you have summarized the feature of borrowing by exempt.  We updated the KEP to describe it, see https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1040-priority-and-fairness#dispatching .

The default number of nominal concurrency shares for the `exempt` priority level is zero, as documented in the comment on this field.

The reason to allow the `exempt` priority level to have a non-zero number of nominal concurrency shares is so that the effect of its borrowing on other priority levels can be limited by a cluster administrator.  By allocating some concurrency to `exempt`, that means it only borrows when it needs more than that.  The reason that the default is zero is to maintain consistency with the previous shares->absolute mapping.

File: pkg/apis/flowcontrol/validation/validation.go [hunk 7]
```diff
@@ -369,7 +391,14 @@ func ValidatePriorityLevelConfigurationSpec(spec *flowcontrol.PriorityLevelConfi
 		if spec.Limited != nil {
 			allErrs = append(allErrs, field.Forbidden(fldPath.Child("limited"), "must be nil if the type is not Limited"))
 		}
+		if spec.Exempt != nil {
+			allErrs = append(allErrs, ValidateExemptPriorityLevelConfiguration(spec.Exempt, fldPath.Child("exempt"))...)
+		}
 	case flowcontrol.PriorityLevelEnablementLimited:
+		if spec.Exempt != nil {
+			allErrs = append(allErrs, field.Required(fldPath.Child("exempt"), "must be nil if the type is Limited"))
```
Comment (deads2k) [inline pkg/apis/flowcontrol/validation/validation.go:399]:
```suggestion
			allErrs = append(allErrs, field.Forbidden(fldPath.Child("exempt"), "must be nil if the type is Limited"))
```

File: staging/src/k8s.io/api/flowcontrol/v1beta1/types.go [hunk 8]
```diff
@@ -505,6 +515,43 @@ type LimitedPriorityLevelConfiguration struct {
...
+	// priority levels in proportion to their NCS values:
+	//
+	// NominalCL(i)  = ceil( ServerCL * NCS(i) / sum_ncs )
+	// sum_ncs = sum[priority level k] NCS(k)
+	//
+	// Bigger numbers mean a larger nominal concurrency limit,
+	// at the expense of every other priority level.
+	// This field has a default value of zero.
+	// +optional
+	NominalConcurrencyShares int32 `json:"nominalConcurrencyShares" protobuf:"varint,1,opt,name=nominalConcurrencyShares"`
```
Comment (deads2k) [inline staging/src/k8s.io/api/flowcontrol/v1beta1/types.go:538]: optional fields need to be pointers.

File: staging/src/k8s.io/api/flowcontrol/v1beta1/types.go [hunk 9]
```diff
@@ -505,6 +515,43 @@ type LimitedPriorityLevelConfiguration struct {
...
+	// but affects the other priority levels through the borrowing mechanism.
+	// The server's concurrency limit (ServerCL) is divided among all the
+	// priority levels in proportion to their NCS values:
+	//
+	// NominalCL(i)  = ceil( ServerCL * NCS(i) / sum_ncs )
+	// sum_ncs = sum[priority level k] NCS(k)
+	//
+	// Bigger numbers mean a larger nominal concurrency limit,
+	// at the expense of every other priority level.
+	// This field has a default value of zero.
```
Comment (deads2k) [inline staging/src/k8s.io/api/flowcontrol/v1beta1/types.go:536]: I don't see this actually defaulted.  It's ok to indicate: "unspecified is equivalent to zero".

File: pkg/apis/flowcontrol/validation/validation.go [hunk 10]
```diff
@@ -399,6 +428,17 @@ func ValidateLimitedPriorityLevelConfiguration(lplc *flowcontrol.LimitedPriority
 	return allErrs
 }
 
+func ValidateExemptPriorityLevelConfiguration(eplc *flowcontrol.ExemptPriorityLevelConfiguration, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if eplc.NominalConcurrencyShares < 0 {
```
Comment (deads2k) [inline pkg/apis/flowcontrol/validation/validation.go:433]: once the optional field is a pointer, this will need a nil check.

**Review by deads2k — CHANGES_REQUESTED** [review]
minor comments and then the API lgtm.  I have not reviewed the impl.

File: staging/src/k8s.io/api/flowcontrol/v1beta1/types.go (hunk 8 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/api/flowcontrol/v1beta1/types.go:538]: I copied what is done in the parallel struct LimitedPriorityLevelConfiguration.
I think that the time to make a change, if any, would be at the next API version (v1beta4 or, hopefully, v1).

File: pkg/apis/flowcontrol/validation/validation.go (hunk 10 above)
Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/validation/validation.go:433]: yes

File: staging/src/k8s.io/api/flowcontrol/v1beta1/types.go (hunk 9 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/api/flowcontrol/v1beta1/types.go:536]: I am not sure I understand what you mean by that field not being "actually defaulted".  When a field is left unspecified, it normally gets its default value.  That is what is going on here too.

File: staging/src/k8s.io/api/flowcontrol/v1beta1/types.go (hunk 8 above)
Comment (wojtek-t) [reply staging/src/k8s.io/api/flowcontrol/v1beta1/types.go:538]: We can't change `LimitedPriorityLevelConfiguration`, but let's actually change this one to pointer now (let's not introduce a new places that are not following the best practices).

File: staging/src/k8s.io/api/flowcontrol/v1beta1/types.go (hunk 9 above)
Comment (wojtek-t) [reply staging/src/k8s.io/api/flowcontrol/v1beta1/types.go:536]: This is true for non-pointers, but once we switch to pointer, the default will effectively be "nil". And then we should probably add a defaulting function, like here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/flowcontrol/v1beta3/defaults.go#L47

**Comment by MikeSpreitzer** [comment]
The force-push to 90d57e6a7e7 makes two changes:
1. Change ExemptPriorityLevelConfiguration.NominalConcurrencyShares from an integer to a pointer to an integer
2. Fix some tests to match the change in validation.

File: staging/src/k8s.io/api/flowcontrol/v1beta1/types.go (hunk 9 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/api/flowcontrol/v1beta1/types.go:536]: Done in the force-push to 90d57e6a7e7

File: pkg/apis/flowcontrol/validation/validation.go (hunk 10 above)
Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/validation/validation.go]: Done in the force-push to 90d57e6a7e7

**Comment by wojtek-t** [comment]
/lgtm

@deads2k - back to you for approval

File: pkg/apis/flowcontrol/v1alpha1/defaults.go [hunk 11]
```diff
@@ -40,6 +40,17 @@ func SetDefaults_FlowSchemaSpec(spec *v1alpha1.FlowSchemaSpec) {
 	}
 }
 
+func SetDefaults_ExemptPriorityLevelConfiguration(eplc *v1alpha1.ExemptPriorityLevelConfiguration) {
+	if *eplc.NominalConcurrencyShares == 0 {
```
Comment (deads2k) [inline pkg/apis/flowcontrol/v1alpha1/defaults.go:44]:
```suggestion
	if eplc.NominalConcurrencyShares == nil {
```

we only want to default on nil, right?

File: pkg/apis/flowcontrol/v1beta1/defaults.go [hunk 12]
```diff
@@ -40,6 +40,17 @@ func SetDefaults_FlowSchemaSpec(spec *v1beta1.FlowSchemaSpec) {
 	}
 }
 
+func SetDefaults_ExemptPriorityLevelConfiguration(eplc *v1beta1.ExemptPriorityLevelConfiguration) {
+	if *eplc.NominalConcurrencyShares == 0 {
```
Comment (deads2k) [inline pkg/apis/flowcontrol/v1beta1/defaults.go:44]:
```suggestion
	if *eplc.NominalConcurrencyShares == nil {
```

File: pkg/apis/flowcontrol/v1beta2/defaults.go [hunk 13]
```diff
@@ -40,6 +40,17 @@ func SetDefaults_FlowSchemaSpec(spec *v1beta2.FlowSchemaSpec) {
 	}
 }
 
+func SetDefaults_ExemptPriorityLevelConfiguration(eplc *v1beta2.ExemptPriorityLevelConfiguration) {
+	if *eplc.NominalConcurrencyShares == 0 {
```
Comment (deads2k) [inline pkg/apis/flowcontrol/v1beta2/defaults.go:44]:
```suggestion
	if *eplc.NominalConcurrencyShares == nil {
```

File: pkg/apis/flowcontrol/v1beta3/defaults.go [hunk 14]
```diff
@@ -40,6 +40,17 @@ func SetDefaults_FlowSchemaSpec(spec *v1beta3.FlowSchemaSpec) {
 	}
 }
 
+func SetDefaults_ExemptPriorityLevelConfiguration(eplc *v1beta3.ExemptPriorityLevelConfiguration) {
+	if *eplc.NominalConcurrencyShares == 0 {
```
Comment (deads2k) [inline pkg/apis/flowcontrol/v1beta3/defaults.go:44]:
```suggestion
	if *eplc.NominalConcurrencyShares == nil {
```

**Comment by deads2k** [comment]
defaulting looks off to me.  lgtm otherwise.

File: pkg/apis/flowcontrol/v1beta1/defaults.go (hunk 12 above)
Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/v1beta1/defaults.go:44]: Oh crud, the asterisk.

**Comment by MikeSpreitzer** [comment]
I fixed the int/pointer bug in the defaulting logic.
But now I need help with a mystery in the generated code.
Although I provided `SetDefaults_ExemptPriorityLevelConfiguration`, the generated code (shown below) does not use it!  What's going on here?

```go
func SetObjectDefaults_PriorityLevelConfiguration(in *v1beta3.PriorityLevelConfiguration) {
	if in.Spec.Limited != nil {
		SetDefaults_LimitedPriorityLevelConfiguration(in.Spec.Limited)
		if in.Spec.Limited.LimitResponse.Queuing != nil {
			SetDefaults_QueuingConfiguration(in.Spec.Limited.LimitResponse.Queuing)
		}
	}
}
```

**Comment by MikeSpreitzer** [comment]
The force-push to 90605a28b95 fixes the int/pointer mistake in the defaulting logic.

**Comment by deads2k** [comment]
reapplying wojtek's impl lgtm and API lgtm

/lgtm
/approve

**Comment by deads2k** [comment]
need to come back after https://github.com/kubernetes/kubernetes/pull/119305 merges

**Comment by MikeSpreitzer** [comment]
Ugh, there is also a problem with round trip checking :-(

**Comment by MikeSpreitzer** [comment]
Forgot to say `,omitempty` on the exempt NominalConcurrencyShares field.  But that is not (all of) it, the test is also failing on the LendablePercent field.

File: pkg/apis/flowcontrol/v1alpha1/defaults.go (hunk 11 above)
Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/v1alpha1/defaults.go]: done.

File: pkg/apis/flowcontrol/v1beta1/defaults.go (hunk 12 above)
Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/v1beta1/defaults.go]: fixed.

File: pkg/apis/flowcontrol/v1beta2/defaults.go (hunk 13 above)
Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/v1beta2/defaults.go]: fixed.

File: pkg/apis/flowcontrol/v1beta3/defaults.go (hunk 14 above)
Comment (MikeSpreitzer) [reply pkg/apis/flowcontrol/v1beta3/defaults.go]: fixed.

**Comment by MikeSpreitzer** [comment]
/unhold
The force-push to ce90eb2cc2f is a rebase onto `master`, after #119305 merged.
